### PR TITLE
ensures no compete on onSubscribe and onComplete calls

### LINKED ISSUE #1684: MonoToCompletableFuture#get never return because the invocation of onComplete method before onSubscribe
The problem comes from [https://github.com/spring-projects/spring-framework/issues/26464#issue-795660549](https://github.com/spring-projects/spring-framework/issues/26464#issue-795660549)

## Steps to Reproduce
A demo to reproduce:[https://github.com/codergmc/websocket-demo](https://github.com/codergmc/websocket-demo)

- rabbitmq(enable stomp):3.8.16 
- java:adopt-openjdk-15.0.2

- Visit [http://localhost:9999/test](http://localhost:9999/test)

- Visit [http://localhost:9999/checking](http://localhost:9999/checking) and problem occured when the result as follow

```
java.base@15.0.2/jdk.internal.misc.Unsafe.park(Native Method)
java.base@15.0.2/java.util.concurrent.locks.LockSupport.park(LockSupport.java:211)
java.base@15.0.2/java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1860) 
java.base@15.0.2/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3137) 
java.base@15.0.2/java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1887) 
java.base@15.0.2/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2062) 
app//reactor.core.publisher.MonoToCompletableFuture.get(MonoToCompletableFuture.java:81) 
app//org.springframework.util.concurrent.CompletableToListenableFutureAdapter.get(CompletableToListenableFutureAdapter.java:99) 
app//org.springframework.messaging.simp.stomp.StompBrokerRelayMessageHandler$SystemSessionConnectionHandler.forward(StompBrokerRelayMessageHandler.java:1081) 
.......
```

- You can find some lastest logs as follow.This mean  the invocation of `MonoToCompletableFuture#onComplete` method before `MonoToCompletableFuture#onSubscribe`.Because `MonoToCompletableFuture.ref` is null when invoke `MonoToCompletableFuture#onComplete` method

```
2021-06-23 05:14:28.062  INFO 37138 --- [ient-loop-nio-2] r.c.publisher.MonoToCompletableFuture    : onComplete this:reactor.core.publisher.MonoToCompletableFuture@379a9335[Not completed] ref:null
2021-06-23 05:14:28.062  INFO 37138 --- [    test_thread] r.c.publisher.MonoToCompletableFuture    : onSubscribe this:reactor.core.publisher.MonoToCompletableFuture@379a9335[Not completed] ref:null
2021-06-23 05:14:28.062  INFO 37138 --- [    test_thread] r.c.publisher.MonoToCompletableFuture    : get start this:reactor.core.publisher.MonoToCompletableFuture@379a9335[Not completed, 1 dependents] ref:reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber@55777635
```

### PR DESCRIPTION
**Pull request by OlegDokuka**
closes #1684 

Signed-off-by: Oleh Dokuka <odokuka@vmware.com>
Signed-off-by: Oleh Dokuka <oleh.dokuka@icloud.com>

### CODE REVIEW

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java [hunk 1]
```diff
@@ -126,20 +113,41 @@ public void subscribe(CoreSubscriber<? super Void> s) {
...
+			return;
+		}
+
+		// add listener to the future to propagate on complete when future is done
+		// addListener likely to be thread safe method
+		// Returned value is deliberately ignored
+		future.addListener(fs);
+
+		// check once again if is cancelled to see if we need to removeListener in case addListener racing with
+		// subscription.cancel (which shell remove listener)
```
Comment (simonbasle) [inline reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:148]: typo: `should`, not `shell`

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java [hunk 2]
```diff
@@ -167,6 +177,11 @@ public Context get() {
 		@Override
 		@SuppressWarnings("FutureReturnValueIgnored")
 		public void cancel() {
+			// cancel is not thread safe since we assume that removeListener is thread-safe. That said if we have
+			// concurrent addListener and removeListener and if addListener is after removeListener, the other Thread
+			// after execution addListener should see changes happened before removeListener. Thus, it shell see
```
Comment (simonbasle) [inline reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:182]: same typo as above: `should`, not `shell`

**Review by simonbasle — COMMENTED** [review]
small typo, but otherwise looks good to me. great to have race tests

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 1 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java]: done

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 2 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java]: done

**Review by simonbasle — APPROVED** [review]

File: reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java [hunk 3]
```diff
@@ -76,4 +89,132 @@ void testDeferredFutureMonoLater() {
...
+		}
+	}
+
+	static class TestSubscriber implements CoreSubscriber<Void> {
+
+		enum Operation {
+			ON_ERROR, ON_COMPLETE, ON_SUBSCRIBE
+		}
+
+		volatile     Subscription                                              s;
```
Comment (hyperxpro) [inline reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:189]: Maybe remove trailing whitespaces?

Comment (violetagg) [reply reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:189]: @hyperxpro I do not see trailing whitespaces, also we have checks for that and I would expect that they will fail.
https://github.com/reactor/reactor-netty/blob/a86addb52836342b807e09345d9fcc99c4e0abb5/codequality/checkstyle.xml#L23-L27

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java [hunk 4]
```diff
@@ -80,21 +80,8 @@
 
 		@Override
 		@SuppressWarnings("FutureReturnValueIgnored")
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:82]: nit: remove the suppression above @SuppressWarnings("FutureReturnValueIgnored")

it is related to `addListener` invocation

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java [hunk 5]
```diff
@@ -126,20 +113,41 @@ public void subscribe(CoreSubscriber<? super Void> s) {
 				return;
 			}
 
-			if (f.isDone()) {
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:129]: nit: remove the suppression above @SuppressWarnings("FutureReturnValueIgnored")

it is related to `addListener` invocation

File: reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java (hunk 3 above)
Comment (hyperxpro) [reply reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:189]: So these whitespaces between `volatile` to `Subscription`, and `Subscription` to `s` are normal as per code style?

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java [hunk 6]
```diff
@@ -126,20 +113,41 @@ public void subscribe(CoreSubscriber<? super Void> s) {
...
-				}
-				else {
-					Operators.error(s, FutureSubscription.wrapError(f.cause()));
-				}
-				return;
+			doSubscribe(s, f);
+		}
+	}
+
+	static <F extends Future<Void>> void doSubscribe(CoreSubscriber<? super Void> s, F future) {
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:120]: nit: add suppression `@SuppressWarnings("FutureReturnValueIgnored")`

it is related to `addListener`/`removeListener` invocation

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java [hunk 7]
```diff
@@ -126,20 +113,41 @@ public void subscribe(CoreSubscriber<? super Void> s) {
...
+
+		// add listener to the future to propagate on complete when future is done
+		// addListener likely to be thread safe method
+		// Returned value is deliberately ignored
+		future.addListener(fs);
+
+		// check once again if is cancelled to see if we need to removeListener in case addListener racing with
+		// subscription.cancel (which should remove listener)
+		if (fs.cancelled) {
+			future.removeListener(fs);
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:150]: nit: Add comment that the returned value is deliberately ignored

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java [hunk 8]
```diff
@@ -149,6 +157,8 @@ public void subscribe(CoreSubscriber<? super Void> s) {
 		final CoreSubscriber<? super Void> s;
 
 		final F                            future;
+
+		boolean cancelled;
 		FutureSubscription(F future, CoreSubscriber<? super Void> s) {
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:162]: nit: add empty row between field and constructor

File: reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java (hunk 3 above)
Comment (simonbasle) [reply reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:189]: the spaces between volatile and Subscription not really. the second set of spaces is usual, to align the volatile variable with its associated `Atomic*FieldUpdater`

Comment (hyperxpro) [reply reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:189]: Alright, that makes sense. I just started reading the source codes so not familiar with the coding style. Thanks for the input @violetagg @simonbasle :)

File: reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java [hunk 9]
```diff
@@ -76,4 +89,132 @@ void testDeferredFutureMonoLater() {
...
+			RaceTestUtils.race(() -> FutureMono.deferFuture(promiseSupplier)
+			                                   .subscribe(subscriber), subscriber::cancel);
+
+			assertThat(resolveListeners(promise)).isNullOrEmpty();
+			assertThat(subscriber.operations).first()
+			                                 .isEqualTo(TestSubscriber.Operation.ON_SUBSCRIBE);
+		}
+	}
+
+	static GenericFutureListener<? extends Future<?>>[] resolveListeners(Promise<Void> promise) {
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:163]: nit: Add `@Nullable`

**Review by violetagg — CHANGES_REQUESTED** [review]
Please resolve the build warning:

```
> Task :reactor-netty-core:compileTestJava
/.../reactor-netty/reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:176: warning: [unchecked] unchecked cast
                        return (GenericFutureListener<? extends Future<?>>[]) listeners1.get(objListener);
                                                                                            ^
  required: GenericFutureListener<? extends Future<?>>[]
  found:    Object
1 warning
```

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 8 above)
Comment (violetagg) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:162]: I have to check how to add this to checkstyle configuration ...

File: reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java (hunk 9 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/test/java/reactor/netty/FutureMonoTest.java:166]: done

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 8 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:163]: done

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 7 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:150]: done

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 6 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:119]: done

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 5 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java:129]: done

File: reactor-netty-core/src/main/java/reactor/netty/FutureMono.java (hunk 4 above)
Comment (OlegDokuka) [reply reactor-netty-core/src/main/java/reactor/netty/FutureMono.java]: done

**Review by violetagg — APPROVED** [review]

**Comment by violetagg** [comment]
@OlegDokuka Thanks for the PR!

**Pull request merged**